### PR TITLE
feat(toolbar-button-clicked) Enhance toolbar buttons with notify click

### PR DESCRIPTION
**Pull request by horymury**
- add possibility to allow execution of the button's routine besides triggering
`toolbarButtonClicked` API event
- keep backwards compatibility

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -585,41 +585,56 @@ var config = {
     //     alwaysVisible: false
     // },
 
-    // Toolbar buttons which have their click event exposed through the API on
-    // `toolbarButtonClicked` event instead of executing the normal click routine.
+    // Toolbar buttons which have their click/tap event exposed through the API on
+    // `toolbarButtonClicked`. Passing a string for the button key will
+    // prevent execution of the click/tap routine; passing an object with `key` and
```
Comment (quitrk) [inline config.js:590]: passing just a string should not prevent execution right? for backwards compatibility

File: modules/API/API.js [hunk 2]
```diff
@@ -1511,12 +1511,14 @@ class API {
      * Notify external application ( if API is enabled) that a toolbar button was clicked.
      *
      * @param {string} key - The key of the toolbar button.
+     * @param {boolean} preventExecution - Whether execution of the button click was prevented or not.
```
Comment (quitrk) [inline modules/API/API.js:1514]: is this a feature request?

File: react/features/base/toolbox/components/AbstractVideoMuteButton.js [hunk 3]
```diff
@@ -22,11 +23,13 @@ export default class AbstractVideoMuteButton<P : Props, S : *>
      * @returns {void}
      */
     _handleClick() {
-        const { handleClick } = this.props;
+        const { handleClick, notifyMode } = this.props;
 
         if (handleClick) {
```
Comment (quitrk) [inline react/features/base/toolbox/components/AbstractVideoMuteButton.js:28]: shouldn't this be prevented as well?

File: react/features/base/toolbox/components/AbstractAudioMuteButton.js [hunk 4]
```diff
@@ -23,11 +24,13 @@ export default class AbstractAudioMuteButton<P: Props, S: *>
      * @returns {void}
      */
     _handleClick() {
-        const { handleClick } = this.props;
+        const { handleClick, notifyMode } = this.props;
 
         if (handleClick) {
             handleClick();
+        }
```
Comment (quitrk) [inline react/features/base/toolbox/components/AbstractAudioMuteButton.js:31]: shouldn't `handleClick` be also prevented from executing?

File: react/features/chat/components/web/ChatButton.js [hunk 5]
```diff
@@ -60,8 +60,6 @@ class ChatButton extends AbstractButton<Props, *> {
 
         if (handleClick) {
             handleClick();
```
Comment (quitrk) [inline react/features/chat/components/web/ChatButton.js:62]: shouldn't this button also take into consideration the execution prevention?

File: react/features/participants-pane/components/web/ParticipantsPaneButton.js [hunk 6]
```diff
@@ -36,8 +36,6 @@ class ParticipantsPaneButton extends AbstractButton<Props, *> {
 
         if (handleClick) {
             handleClick();
-
-            return;
```
Comment (quitrk) [inline react/features/participants-pane/components/web/ParticipantsPaneButton.js:40]: same question

File: react/features/toolbox/components/web/ShareDesktopButton.js [hunk 7]
```diff
@@ -78,8 +78,6 @@ class ShareDesktopButton extends AbstractButton<Props, *> {
 
         if (handleClick) {
             handleClick();
-
-            return;
```
Comment (quitrk) [inline react/features/toolbox/components/web/ShareDesktopButton.js:82]: same question

**Comment by quitrk** [comment]
looks like some of the buttons do not add the prevent execution guard. Also, it seems to me that it would be cleaner to guard in the parent class and call super instead of checking every button

File: react/features/base/toolbox/components/AbstractAudioMuteButton.js (hunk 4 above)
Comment (horymury) [reply react/features/base/toolbox/components/AbstractAudioMuteButton.js:31]: handleClick only does the API notify part ( if defined)

File: react/features/chat/components/web/ChatButton.js (hunk 5 above)
Comment (horymury) [reply react/features/chat/components/web/ChatButton.js:62]: this only does API notify

File: modules/API/API.js (hunk 2 above)
Comment (horymury) [reply modules/API/API.js:1514]: discussed offline

**Comment by horymury** [comment]
> looks like some of the buttons do not add the prevent execution guard. Also, it seems to me that it would be cleaner to guard in the parent class and call super instead of checking every button

We discussed offline, there does not seem to be a simple solution to move the logic on the abstract level and keep refactors at minimum.

File: react/features/toolbox/components/web/ToolbarButton.js [hunk 8]
```diff
@@ -4,6 +4,7 @@ import React from 'react';
```
Comment (quitrk) [inline react/features/toolbox/components/web/ToolbarButton.js:4]: seems like this component is only being used by ReactionsMenuButton and OverflowMenuButton. Can't those just inherit the AbstractButton instead?

Comment (horymury) [reply react/features/toolbox/components/web/ToolbarButton.js]: Yep removed ToolbarButton.js completely and switched to using AbstractButton for the 2.

File: react/features/base/toolbox/components/ToolboxItem.web.js [hunk 9]
```diff
@@ -6,7 +6,15 @@ import { Icon } from '../../icons';
 import { Tooltip } from '../../tooltip';
 
 import AbstractToolboxItem from './AbstractToolboxItem';
-import type { Props } from './AbstractToolboxItem';
+import type { Props as AbastractToolboxItemProps } from './AbstractToolboxItem';
```
Comment (quitrk) [inline react/features/base/toolbox/components/ToolboxItem.web.js:9]: there's a typo here

Comment (horymury) [reply react/features/base/toolbox/components/ToolboxItem.web.js]: fixed

**Review by quitrk — APPROVED** [review]

**Pull request merged**